An Internal Server Error occurs if the request URI has two forward slashes after the crate name

An unexpected `HTTP 500 Internal Server Error` response is returned if the request URI has to forward slashes after the crate name, for example:

`https://docs.rs/winapi//`

![image](https://user-images.githubusercontent.com/39023054/120091872-74b53600-c141-11eb-946a-acb8d205ebfa.png)

Tested on both Google Chrome and Firefox on Windows 21H1.

Thanks to Legorooj#2455 for bringing this up in the Rust Programming Language Community Server.

thank you for the report! 

the actual error when I reproduce this locally is: 
```
[ERROR] docs_rs::web: No error page for status Some(MovedPermanently); The request had a trailing slash.
```

--- a/src/web/mod.rs
+++ b/src/web/mod.rs
@@ -167,18 +167,6 @@ impl Handler for MainHandler {
             }
         }
 
-        fn pass_iron_errors_with_redirect(e: IronError) -> IronResult<Response> {
-            // in some cases the iron router will return a redirect as an `IronError`.
-            // Here we convert these into an `Ok(Response)`.
-            if e.error.downcast_ref::<TrailingSlash>().is_some()
-                || e.response.status == Some(status::MovedPermanently)
-            {
-                Ok(e.response)
-            } else {
-                Err(e)
-            }
-        }
-
         // This is kind of a mess.
         //
         // Almost all files should be served through the `router_handler`; eventually
@@ -195,7 +183,17 @@ impl Handler for MainHandler {
         self.shared_resource_handler
             .handle(req)
             .or_else(|e| if_404(e, || self.router_handler.handle(req)))
-            .or_else(pass_iron_errors_with_redirect)
+            .or_else(|e| {
+                // in some cases the iron router will return a redirect as an `IronError`.
+                // Here we convert these into an `Ok(Response)`.
+                if e.error.downcast_ref::<TrailingSlash>().is_some()
+                    || e.response.status == Some(status::MovedPermanently)
+                {
+                    Ok(e.response)
+                } else {
+                    Err(e)
+                }
+            })
             .or_else(|e| {
                 let err = if let Some(err) = e.error.downcast_ref::<error::Nope>() {
                     *err

--- a/src/web/mod.rs
+++ b/src/web/mod.rs
@@ -109,7 +109,7 @@ use iron::{
 };
 use page::TemplateData;
 use postgres::Client;
-use router::NoRoute;
+use router::{NoRoute, TrailingSlash};
 use semver::{Version, VersionReq};
 use serde::Serialize;
 use std::{borrow::Cow, fmt, net::SocketAddr, sync::Arc};
@@ -167,6 +167,18 @@ impl Handler for MainHandler {
             }
         }
 
+        fn pass_iron_errors_with_redirect(e: IronError) -> IronResult<Response> {
+            // in some cases the iron router will return a redirect as an `IronError`.
+            // Here we convert these into an `Ok(Response)`.
+            if e.error.downcast_ref::<TrailingSlash>().is_some()
+                || e.response.status == Some(status::MovedPermanently)
+            {
+                Ok(e.response)
+            } else {
+                Err(e)
+            }
+        }
+
         // This is kind of a mess.
         //
         // Almost all files should be served through the `router_handler`; eventually
@@ -183,6 +195,7 @@ impl Handler for MainHandler {
         self.shared_resource_handler
             .handle(req)
             .or_else(|e| if_404(e, || self.router_handler.handle(req)))
+            .or_else(pass_iron_errors_with_redirect)
             .or_else(|e| {
                 let err = if let Some(err) = e.error.downcast_ref::<error::Nope>() {
                     *err
@@ -736,6 +749,20 @@ mod test {
             // with or without slash
             assert_redirect("/proc-macro", target, web)?;
             assert_redirect("/proc-macro/", target, web)?;
+            Ok(())
+        })
+    }
+
+    #[test]
+    fn double_slash_does_redirect_and_remove_slash() {
+        wrapper(|env| {
+            env.fake_release()
+                .name("bat")
+                .version("0.2.0")
+                .create()
+                .unwrap();
+            let web = env.frontend();
+            assert_redirect("/bat//", "/bat/0.2.0/bat/", web)?;
             Ok(())
         })
     }

